swarm/runtime/service.py

```
@@ -298,41 +300,65 @@ def list_runs_paginated(
+        active_ids.sort()
+        legacy_ids.sort()
+
+        # 2. Compute Total (Optimistic)
+        # We perform a fast calculation assuming disjoint sets for speed.
+        # This gives O(1) time complexity for total count instead of O(N) deduplication.
+        # Overestimate is rare and documented as acceptable.
+        total = len(example_ids) + len(active_ids) + len(legacy_ids)
+
+        # 3. Stream and Slice using Generators (O(limit) instead of O(N))
+        # Use heapq.merge to combine sorted lists without concatenation or full sort
+        def stream_unique():
```

Ruff (0.14.14)

Missing return type annotation for private function `stream_unique`

(ANN202)